File: api/src/main/java/org/pmiops/workbench/reporting/ReportingUploadServiceStreamingImpl.java
package org.pmiops.workbench.reporting;

import static org.pmiops.workbench.reporting.insertion.ColumnValueExtractorUtils.getBigQueryTableName;

import com.google.cloud.bigquery.BigQueryError;
import com.google.cloud.bigquery.InsertAllRequest;
import com.google.cloud.bigquery.InsertAllResponse;
import com.google.cloud.bigquery.TableId;
import com.google.common.base.Stopwatch;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Logger;
import javax.inject.Provider;
import org.pmiops.workbench.api.BigQueryService;
import org.pmiops.workbench.config.WorkbenchConfig;
import org.pmiops.workbench.model.ReportingCohort;
import org.pmiops.workbench.model.ReportingDataset;
import org.pmiops.workbench.model.ReportingDatasetCohort;
import org.pmiops.workbench.model.ReportingDatasetConceptSet;
import org.pmiops.workbench.model.ReportingDatasetDomainIdValue;
import org.pmiops.workbench.model.ReportingInstitution;
import org.pmiops.workbench.model.ReportingSnapshot;
import org.pmiops.workbench.model.ReportingUser;
import org.pmiops.workbench.model.ReportingWorkspace;
import org.pmiops.workbench.reporting.insertion.CohortColumnValueExtractor;
import org.pmiops.workbench.reporting.insertion.ColumnValueExtractor;
import org.pmiops.workbench.reporting.insertion.DatasetCohortColumnValueExtractor;
import org.pmiops.workbench.reporting.insertion.DatasetColumnValueExtractor;
import org.pmiops.workbench.reporting.insertion.DatasetConceptSetColumnValueExtractor;
import org.pmiops.workbench.reporting.insertion.DatasetDomainColumnValueExtractor;
import org.pmiops.workbench.reporting.insertion.InsertAllRequestPayloadTransformer;
import org.pmiops.workbench.reporting.insertion.InstitutionColumnValueExtractor;
import org.pmiops.workbench.reporting.insertion.UserColumnValueExtractor;
import org.pmiops.workbench.reporting.insertion.WorkspaceColumnValueExtractor;
import org.pmiops.workbench.utils.LogFormatters;
import org.springframework.stereotype.Service;

@Service
public class ReportingUploadServiceStreamingImpl implements ReportingUploadService {
  private static final Logger log =
      Logger.getLogger(ReportingUploadServiceStreamingImpl.class.getName());
  private static final InsertAllRequestPayloadTransformer<ReportingCohort> cohortRequestBuilder =
      CohortColumnValueExtractor::values;
  private static final InsertAllRequestPayloadTransformer<ReportingDatasetCohort>
      datasetCohortRequestBuilder = DatasetCohortColumnValueExtractor::values;
  private static final InsertAllRequestPayloadTransformer<ReportingDatasetConceptSet>
      datasetConceptSetRequestBuilder = DatasetConceptSetColumnValueExtractor::values;
  private static final InsertAllRequestPayloadTransformer<ReportingDatasetDomainIdValue>
      datasetDomainIIdValueRequestBuilder = DatasetDomainColumnValueExtractor::values;
  private static final InsertAllRequestPayloadTransformer<ReportingInstitution>
      institutionRequestBuilder = InstitutionColumnValueExtractor::values;
  private static final InsertAllRequestPayloadTransformer<ReportingUser> userRequestBuilder =
      UserColumnValueExtractor::values;
  private static final InsertAllRequestPayloadTransformer<ReportingWorkspace>
      workspaceRequestBuilder = WorkspaceColumnValueExtractor::values;
  private static final InsertAllRequestPayloadTransformer<ReportingDataset> datasetRequestBuilder =
      DatasetColumnValueExtractor::values;

  private final BigQueryService bigQueryService;
  private final ReportingVerificationService reportingVerificationService;
  private final Provider<WorkbenchConfig> configProvider;
  private final Provider<Stopwatch> stopwatchProvider;

  public ReportingUploadServiceStreamingImpl(
      BigQueryService bigQueryService,
      ReportingVerificationService reportingVerificationService,
      Provider<WorkbenchConfig> configProvider,
      Provider<Stopwatch> stopwatchProvider) {
    this.bigQueryService = bigQueryService;
    this.reportingVerificationService = reportingVerificationService;
    this.configProvider = configProvider;
    this.stopwatchProvider = stopwatchProvider;
  }

  @Override
  public void uploadSnapshot(ReportingSnapshot reportingSnapshot) {
    final Stopwatch stopwatch = stopwatchProvider.get();
    final ImmutableMap.Builder<TableId, InsertAllResponse> responseMapBuilder =
        ImmutableMap.builder();
    final List<InsertAllRequest> insertAllRequests = getInsertAllRequests(reportingSnapshot);
    final StringBuilder performanceStringBuilder = new StringBuilder();
    for (InsertAllRequest request : insertAllRequests) {
      stopwatch.start();
      final InsertAllResponse currentResponse = bigQueryService.insertAll(request);
      responseMapBuilder.put(request.getTable(), currentResponse);
      stopwatch.stop();
      performanceStringBuilder.append(
          LogFormatters.rate(
                  String.format("Streaming upload into %s table", request.getTable().getTable()),
                  stopwatch.elapsed(),
                  request.getRows().size(),
                  "rows")
              + "\n");
      stopwatch.reset();
    }
    log.info(performanceStringBuilder.toString());
    checkResponseAndRowCounts(reportingSnapshot, responseMapBuilder.build());
  }

  private void checkResponseAndRowCounts(
      ReportingSnapshot reportingSnapshot, ImmutableMap<TableId, InsertAllResponse> responseMap) {
    checkResponse(responseMap);

    reportingVerificationService.verifyAndLog(reportingSnapshot);
  }

  private TableId getTableId(String tableName) {
    final String projectId = configProvider.get().server.projectId;
    final String dataset = configProvider.get().reporting.dataset;
    return TableId.of(projectId, dataset, tableName);
  }

  private <E extends Enum<E> & ColumnValueExtractor<?>> TableId getTableId(
      Class<E> columnValueExtractorClass) {
    final String projectId = configProvider.get().server.projectId;
    final String dataset = configProvider.get().reporting.dataset;

    return TableId.of(projectId, dataset, getBigQueryTableName(columnValueExtractorClass));
  }

  private List<InsertAllRequest> getInsertAllRequests(ReportingSnapshot reportingSnapshot) {
    final Map<String, Object> fixedValues =
        ImmutableMap.of("snapshot_timestamp", reportingSnapshot.getCaptureTimestamp());

    return ImmutableList.of(
            cohortRequestBuilder.build(
                getTableId(CohortColumnValueExtractor.class),
                reportingSnapshot.getCohorts(),
                fixedValues),
            datasetRequestBuilder.build(
                getTableId(DatasetColumnValueExtractor.class),
                reportingSnapshot.getDatasets(),
                fixedValues),
            datasetCohortRequestBuilder.build(
                getTableId(DatasetCohortColumnValueExtractor.class),
                reportingSnapshot.getDatasetCohorts(),
                fixedValues),
            datasetConceptSetRequestBuilder.build(
                getTableId(DatasetConceptSetColumnValueExtractor.class),
                reportingSnapshot.getDatasetConceptSets(),
                fixedValues),
            datasetDomainIIdValueRequestBuilder.build(
                getTableId(DatasetDomainColumnValueExtractor.class),
                reportingSnapshot.getDatasetDomainIdValues(),
                fixedValues),
            institutionRequestBuilder.build(
                getTableId(InstitutionColumnValueExtractor.class),
                reportingSnapshot.getInstitutions(),
                fixedValues),
            userRequestBuilder.build(
                getTableId(UserColumnValueExtractor.class),
                reportingSnapshot.getUsers(),
                fixedValues),
            workspaceRequestBuilder.build(
                getTableId(WorkspaceColumnValueExtractor.class),
                reportingSnapshot.getWorkspaces(),
                fixedValues))
        .stream()
        .filter(r -> r.getRows().size() > 0)
        .collect(ImmutableList.toImmutableList());
  }

  private void checkResponse(Map<TableId, InsertAllResponse> tableIdToResponse) {
    final boolean anyErrors =
        tableIdToResponse.values().stream().anyMatch(InsertAllResponse::hasErrors);

    if (anyErrors) {
      for (Map.Entry<TableId, InsertAllResponse> entry : tableIdToResponse.entrySet()) {
        final TableId tableId = entry.getKey();
        final InsertAllResponse response = entry.getValue();
        if (response.hasErrors()) {
          final Map<Long, List<BigQueryError>> rowIdToErrors = response.getInsertErrors();
          rowIdToErrors.forEach(
              (rowId, errors) -> {
                log.severe(
                    String.format(
                        "%d errors inserting row %d in table %s",
                        errors.size(), rowId, tableId.getTable()));
                errors.forEach(
                    error ->
                        log.severe(
                            String.format(
                                "  %s at %s: %s",
                                error.getReason(), error.getLocation(), error.getMessage())));
              });
        }
      }
      throw new RuntimeException(
          String.format(
              "Encountered errors in %d tables in upload to BigQuery.",
              tableIdToResponse.keySet().size()));
    }
  }
}

File: api/src/main/java/org/pmiops/workbench/reporting/ReportingUploadServiceStreamingImpl.kt
package org.pmiops.workbench.reporting

import com.google.cloud.bigquery.BigQueryError
import com.google.cloud.bigquery.InsertAllRequest
import com.google.cloud.bigquery.InsertAllResponse
import com.google.cloud.bigquery.TableId
import com.google.common.base.Stopwatch
import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableMap
import org.pmiops.workbench.api.BigQueryService
import org.pmiops.workbench.config.WorkbenchConfig
import org.pmiops.workbench.model.ReportingSnapshot
import org.pmiops.workbench.reporting.ReportingUploadServiceStreamingImpl
import org.pmiops.workbench.reporting.insertion.CohortColumnValueExtractor
import org.pmiops.workbench.reporting.insertion.InsertAllRequestPayloadTransformer
import org.pmiops.workbench.reporting.insertion.InstitutionColumnValueExtractor
import org.pmiops.workbench.reporting.insertion.UserColumnValueExtractor
import org.pmiops.workbench.reporting.insertion.WorkspaceColumnValueExtractor
import org.pmiops.workbench.utils.LogFormatters
import org.springframework.stereotype.Service
import java.util.function.Consumer
import java.util.logging.Logger
import javax.inject.Provider

@Service("REPORTING_UPLOAD_SERVICE_STREAMING_IMPL")
class ReportingUploadServiceStreamingImpl(
        private val bigQueryService: BigQueryService,
        private val configProvider: Provider<WorkbenchConfig>,
        private val stopwatchProvider: Provider<Stopwatch>) : ReportingUploadService {
    override fun uploadSnapshot(reportingSnapshot: ReportingSnapshot) {
        val stopwatch = stopwatchProvider.get()
        val responseMapBuilder = ImmutableMap.builder<TableId, InsertAllResponse>()
        val insertAllRequests = getInsertAllRequests(reportingSnapshot)
        for (request in insertAllRequests) {
            stopwatch.start()
            val currentResponse = bigQueryService.insertAll(request)
            responseMapBuilder.put(request.table, currentResponse)
            stopwatch.stop()
            log.info(
                    LogFormatters.rate(String.format("Streaming upload into %s table", request.table.table),
                            stopwatch.elapsed(),
                            request.rows.size.toDouble(),
                            "rows"))
            stopwatch.reset()
        }
        checkResults(responseMapBuilder.build())
    }

    private fun getTableId(tableName: String): TableId {
        val projectId = configProvider.get().server.projectId
        val dataset = configProvider.get().reporting.dataset
        return TableId.of(projectId, dataset, tableName)
    }

    private fun getInsertAllRequests(reportingSnapshot: ReportingSnapshot): List<InsertAllRequest> {
        val fixedValues: Map<String, Any> = ImmutableMap.of<String, Any>("snapshot_timestamp", reportingSnapshot.captureTimestamp)
        return ImmutableList.of(
                cohortRequestBuilder.build(
                        getTableId(CohortColumnValueExtractor.TABLE_NAME),
                        reportingSnapshot.cohorts,
                        fixedValues),
                institutionRequestBuilder.build(
                        getTableId(InstitutionColumnValueExtractor.TABLE_NAME),
                        reportingSnapshot.institutions,
                        fixedValues),
                userRequestBuilder.build(
                        getTableId(UserColumnValueExtractor.TABLE_NAME),
                        reportingSnapshot.users,
                        fixedValues),
                workspaceRequestBuilder.build(
                        getTableId(WorkspaceColumnValueExtractor.TABLE_NAME),
                        reportingSnapshot.workspaces,
                        fixedValues))
                .stream()
                .filter { r: InsertAllRequest -> r.rows.size > 0 }
                .collect(ImmutableList.toImmutableList())
    }

    private fun checkResults(tableIdToResponse: Map<TableId, InsertAllResponse>) {
        val anyErrors = tableIdToResponse.values.stream().anyMatch { obj: InsertAllResponse -> obj.hasErrors() }
        if (anyErrors) {
            for ((tableId, response) in tableIdToResponse) {
                if (response.hasErrors()) {
                    val rowIdToErrors = response.insertErrors
                    rowIdToErrors.forEach { (rowId: Long?, errors: List<BigQueryError>) ->
                        log.severe(String.format(
                                "%d errors inserting row %d in table %s",
                                errors.size, rowId, tableId.table))
                        errors.forEach(
                                Consumer { error: BigQueryError ->
                                    log.severe(String.format(
                                            "  %s at %s: %s",
                                            error.reason, error.location, error.message))
                                })
                    }
                }
            }
            throw RuntimeException(String.format(
                    "Encountered errors in %d tables in upload to BigQuery.",
                    tableIdToResponse.keys.size))
        }
    }

    companion object {
        private val cohortRequestBuilder = InsertAllRequestPayloadTransformer { CohortColumnValueExtractor.values() }
        private val institutionRequestBuilder = InsertAllRequestPayloadTransformer { InstitutionColumnValueExtractor.values() }
        private val log = Logger.getLogger(ReportingUploadServiceStreamingImpl::class.java.name)
        private val userRequestBuilder = InsertAllRequestPayloadTransformer { UserColumnValueExtractor.values() }
        private val workspaceRequestBuilder = InsertAllRequestPayloadTransformer { WorkspaceColumnValueExtractor.values() }
    }
}
